src/main/java/slimeknights/tconstruct/smeltery/recipe/AbstractCastingRecipe.java

```diff
@@ -0,0 +1,202 @@
+package slimeknights.tconstruct.smeltery.recipe;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.inventory.IInventory;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipe;
+import net.minecraft.item.crafting.IRecipeSerializer;
+import net.minecraft.item.crafting.IRecipeType;
+import net.minecraft.item.crafting.Ingredient;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.JSONUtils;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.world.World;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.registries.ForgeRegistries;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public abstract class AbstractCastingRecipe implements IRecipe<IInventory> {
+  protected final IRecipeType<?> type;
+  protected final ResourceLocation id;
+  protected final String group;
+  private final Ingredient cast;
+  private final FluidStack fluid;
+  private final ItemStack result;
+  protected final int coolingTime;
+  private final boolean consumed;
+  private final boolean switchSlots;
```

Comment:
Some javadoc comments on these fields would be good, explain what they do. Especially the booleans.

You can also use `@Getter` on these to have Lombok make the getter methods for you.